### ISSUE TITLE
Wrong logic in Platform.wait_for_work, and simplify it with property/decorator attribute

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1472]
Hello @dbeltrankyl,

Reading this function for !475 , I realized that the docs and code could probably be improved. For 4.2.0 or later, I think, as it has low priority.

It's late in the day, so my logic may be a bit flawed, but I think the current logic has a few issues;

https://earth.bsc.es/gitlab/es/autosubmit/-/blob/02c71e4d522fe5ede77fe83c5caebdf05f1827dd/autosubmit/platforms/platform.py#L861-868

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/005e466188fbf50a7b3f05176e692a5a/image.png)

The `process_log` starts set at `False`, while this could actually be wrong... I mean, maybe there are already logs. But the process will `sleep` for at least one second anyway, even if unnecessary.

Then, if `self.work_event` is set, or `self.recovery_queue` is not empty, without the `cleanup_event` the loop will proceed to `sleep` one second until the `range` is exhausted. Even if `process_log` was already set to `True`. It would be better to leave earlier.

In a few lines below, there is a while loop on timeout being greather than `0`, and on the `process_log` being `True` (it appears to be the same conditions that some lines above set `process_log` to `True`). So, if I am reading the boolean logic correct, it will sleep one second until the timeout is exhausted or while there are logs to process.

I think the intention was to sleep one second as long as the timeout was greater than zero, and there were **no** logs to process (i.e. inverted value?).

https://earth.bsc.es/gitlab/es/autosubmit/-/blob/02c71e4d522fe5ede77fe83c5caebdf05f1827dd/autosubmit/platforms/platform.py#L871

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/e78687122c9ab513f1f143bd48084b80/image.png)

The `wait_for_work` could have less if's, and use `@property`s to simplify reading the logic, IMHO. Assuming the condition to process logs is isolated, I think one possible direction to improve it could be something like (not tested, would need more :coffee: and write some tests to verify this change):

```diff
diff --git a/autosubmit/platforms/platform.py b/autosubmit/platforms/platform.py
index 9af2ca48..28bd98dd 100644
--- a/autosubmit/platforms/platform.py
+++ b/autosubmit/platforms/platform.py
@@ -854,27 +854,40 @@ class Platform(object):
             self.cleanup_event.set()
             self.log_recovery_process.join()
 
-    def wait_for_work(self, sleep_time=60):
+    @property
+    def process_log(self):
+        """Whether to process logs or not.
+
+        Returns True if the ``work_event`` was set (set in ``job.update_parameters``).
+        Returns True if the there are items in ``recovery_queue`` (populated by ``self.add_job_to_log_recover``).
+        Returns True if the ``cleanup_event`` was set (set at the atexit handler function).
+
+        :returns: True if there are logs waiting to be processed. False otherwise.
         """
-        This function waits for the work_event to be set or the cleanup_event to be set.
+        return self.work_event.is_set() or not self.recovery_queue.empty() or self.cleanup_event.is_set()
+
+    def wait_for_work(self, sleep_time: int = 1) -> bool:
         """
-        process_log = False
-        for remaining in range(sleep_time, 0, -1):  # Min time to wait unless clean-up signal is set
-            time.sleep(1)
-            if self.work_event.is_set() or not self.recovery_queue.empty():
-                process_log = True
-            if self.cleanup_event.is_set():
-                process_log = True
-                break
-        if not process_log: # If no work, wait until the keep_alive_timeout is reached or any signal is set to end the process.
-            timeout = self.keep_alive_timeout - sleep_time
-            while timeout > 0 or not self.recovery_queue.empty() or self.cleanup_event.is_set() or self.work_event.is_set():
-                time.sleep(1)
-                timeout -= 1
-            if not self.recovery_queue.empty() or self.cleanup_event.is_set() or self.work_event.is_set():
-                process_log = True
+        Waits until there are logs to be processed, or until the keep alive timeout is reached.
+
+        It will check he ``self.process_log`` state, and until it is True it will sleep for one second,
+        then check again.
+
+        At the end it will clear the ``self.work_event`` event, and return the True if there are logs
+        to be processed, and False otherwise.
+
+        :param sleep_time: sleep time in seconds.
+        :type sleep_time: int
+        :return: True if there are logs to be processed, False otherwise.
+        :rtype: bool
+        """
+        timeout = self.keep_alive_timeout
+        while timeout > 0 and not self.process_log:
+            time.sleep(sleep_time)
+            timeout -= 1
+
         self.work_event.clear()
-        return process_log
+        return self.process_log
 
     def recover_job_log(self, identifier, jobs_pending_to_process):
         job = None
@@ -913,10 +926,13 @@ class Platform(object):
                 pass
         return jobs_pending_to_process
 
-    def recover_platform_job_logs(self):
+    def recover_platform_job_logs(self) -> None:
         """
         This function, recovers the logs of the jobs that have been submitted.
-        The exit of this process is controlled by the work_event and cleanup_events of the main process.
+
+        See ``self.process_log`` for the property that controls whether there are logs to be processed,
+        or not. And ``self.wait_for_work`` for the function that loops checking that property limited by
+        a timeout.
         """
         setproctitle.setproctitle(f"autosubmit log {self.expid} recovery {self.name.lower()}")
         identifier = f"{self.name.lower()}(log_recovery):"
@@ -928,7 +944,7 @@ class Platform(object):
         Log.result(f"{identifier} Sucessfully connected.")
         log_recovery_timeout = self.config.get("LOG_RECOVERY_TIMEOUT", 60)
         self.keep_alive_timeout = max(log_recovery_timeout*5, 60*5)
-        while self.wait_for_work(sleep_time=max(log_recovery_timeout, 60)):
+        while self.wait_for_work():
             jobs_pending_to_process = self.recover_job_log(identifier, jobs_pending_to_process)
             if self.cleanup_event.is_set():  # Check if main process is waiting for this child to end.
                 self.recover_job_log(identifier, jobs_pending_to_process)
```

If the logic above is correct, I think it would be a bit more readable, easier to maintain, and perform slightly better (but again, I need more :coffee: before saying anything for sure).

Cheers

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 27, 2024, 16:02

Hello @kinow 

Quicky comment about this: 

> Then, if `self.work_event` is set, or `self.recovery_queue` is not empty, without the `cleanup_event` the loop will proceed to `sleep` one second until the `range` is exhausted. Even if `process_log` was already set to `True`. It would be better to leave earlier.

This was intentional. I think I'll not be able to find the comment (since the threads are resolved ), but it was commented somewhere in the !475. 

Since there were complaints about the climate-dt machine's performance, the log process will wait for at least the sleep time specified (I don't know what is set right now) before *trying* to doing work.

So, the idea was to always wait for the sleep time. The only skip for that it is the cleanup is set ( no more jobs to run ) 

Once the wait_for_work function ends, all the jobs in the queue will be processed and wait for more

So, in summary, we could delete that part without touching anything else, and the log would be faster. 

Edit:

Why can it be useful? 

* If there are issues on a platform, the recovery of the platform will be triggered at cpu_clock with a possible ban as there will be always logs to process

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 27, 2024, 16:16

> I think the intention was to sleep one second as long as the timeout was greater than zero, and there were **no** logs to process (i.e. inverted value?).

The intention was to do an active wait. 

There are two timers: 

* A shorter mandatory sleep timer, as commented in the previous comment. ( first for ) 

* The long timeout is set to 5 minutes by default. 

This is the maximum amount of time that the log process will wait before ending itself. 

The intention is to make an active wait. So sleep 1 is to keep trying until the timeout value or whatever the main process indicates

The conditions to avoid the process ends itself are basically that: 

* There is still timeout left
* The recovery_queue is not empty (i.e, logs to process )
* The cleanup is set. Main process is waiting for the process to finish 
* The work_event is set: edit ( The main process is still alive and doing stuff)

That is translated to waiting until there is some log to process because during the mandatory sleeptime it could be that there would still be no log or not work_flag set

**Comment by dbeltrankyl** [comment]
I'm closing this one as the branch was merged, and I think we discussed about this in one AS meeting
